Modular Insurance Eligibility

RobertScalfani opened the pull request:
This PR adds custom modular eligibility inputs to Synthea that allow users to easily dictate eligibility criteria for different insurance companies.

## Input Modular Eligibility File 

There is a new csv eligibilities input file (/resources/payers/insurance_eligibilities.csv) in which you can dictate criteria. It's important to note that changing the columns in this file is not just changing some constants in the code, it's actually creating a new algorithm for each row.

This is an example of what the eligibility input file looks like when defining Medicare/Medicaid/Dual Eligible. The csv file allowed for the removal of baked-in Medicare/Medicaid from Synthea. Their eligibilities are completely defined within this file:

<html xmlns:m="http://schemas.microsoft.com/office/2004/12/omml"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=PowerPoint.Slide>
<meta name=Generator content="Microsoft PowerPoint 15">
<style>
<!--tr
	{mso-height-source:auto;}
col
	{mso-width-source:auto;}
td
	{padding-top:1.0px;
	padding-right:1.0px;
	padding-left:1.0px;
	mso-ignore:padding;
	color:windowtext;
	font-size:18.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:Arial;
	mso-generic-font-family:auto;
	mso-font-charset:0;
	text-align:general;
	vertical-align:bottom;
	border:none;
	mso-background-source:auto;
	mso-pattern:auto;}
.oa1
	{border-top:1.0pt solid #0B2338;
	border-right:1.0pt solid #0B2338;
	border-bottom:3.0pt solid #0B2338;
	border-left:1.0pt solid #0B2338;
	background:#005B93;
	mso-pattern:auto none;
	vertical-align:top;
	padding-bottom:3.6pt;
	padding-left:7.2pt;
	padding-top:3.6pt;
	padding-right:7.2pt;}
.oa2
	{border-top:3.0pt solid #0B2338;
	border-right:1.0pt solid #0B2338;
	border-bottom:1.0pt solid #0B2338;
	border-left:1.0pt solid #0B2338;
	background:#CBD2DC;
	mso-pattern:auto none;
	vertical-align:top;
	padding-bottom:3.6pt;
	padding-left:7.2pt;
	padding-top:3.6pt;
	padding-right:7.2pt;}
.oa3
	{border:1.0pt solid #0B2338;
	background:#E7EAEE;
	mso-pattern:auto none;
	vertical-align:top;
	padding-bottom:3.6pt;
	padding-left:7.2pt;
	padding-top:3.6pt;
	padding-right:7.2pt;}
.oa4
	{border:1.0pt solid #0B2338;
	background:#CBD2DC;
	mso-pattern:auto none;
	vertical-align:top;
	padding-bottom:3.6pt;
	padding-left:7.2pt;
	padding-top:3.6pt;
	padding-right:7.2pt;}
-->
</style>
</head>

<body>
<!--StartFragment-->

Name | … | Age Threshold | Qualifying Codes | Qualifying Attributes | Poverty Multiplier File | Spenddown File | Acceptance Likelihood | Sub-Eligibilities | Logical Operator
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
Social Security Disability |   |   | ssd_eligibility.csv |   |   |   | 0.9 |   |
Standard Medicare |   | 65 |   | end_stage_renal_disease = true |   |   |   | Social Security Disability | or
Standard Medicaid |   |   |  | blindness = true | Payers/ medicaid_income_eligibilities.csv | Payers/ medicaid_mnil_eligibilities.csv |   |   | or
Standard Dual Eligible |   |   |   |   |   |   |   | Standard Medicare \| Standard Medicaid | and

<!--EndFragment-->
</body>

</html>

_insurance_eligibilities.csv_

You can modify existing eligibility criteria or create your own. If you wanted to slightly modify Medicaid to use its same eligibility criteria plus add a new minimum income multiplier threshold of 2.0, you could create a new row where the `Sub Eligibility` is `Standard Medicaid` and the `Poverty Multiplier` is `2.0` with the `Logical Operator` set to `or`. That's all it takes! Similarly to this, you can see how Medicare was defined using `Social Security Disability` as a `Sub Eligibility`. Again, this was used when defining `Dual Eligible` by and-ing the `StandardMedicare` and `StandardMedicaid` eligibility algorithms.
Using the `Sub Eiligibilty` column also allows for complex nesting of and/or logic on top of reusing existing logic.
See **Current Eligibility Column Options** for more details one what each column attribute dictates.

The way this system is set up, it makes it easy for a programmer to extend the column types available. However, a better goal would be feature-completeness to where that would not be necessary (which it may be at/almost at).

_This is the Insurance Companies input file, where the eligibilities can be defined using `Eligibility Algorithms`:_
<html xmlns:m="http://schemas.microsoft.com/office/2004/12/omml"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=PowerPoint.Slide>
<meta name=Generator content="Microsoft PowerPoint 15">
<style>
<!--tr
	{mso-height-source:auto;}
col
	{mso-width-source:auto;}
td
	{padding-top:1.0px;
	padding-right:1.0px;
	padding-left:1.0px;
	mso-ignore:padding;
	color:windowtext;
	font-size:18.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:Arial;
	mso-generic-font-family:auto;
	mso-font-charset:0;
	text-align:general;
	vertical-align:bottom;
	border:none;
	mso-background-source:auto;
	mso-pattern:auto;}
.oa1
	{border-top:1.0pt solid #0B2338;
	border-right:1.0pt solid #0B2338;
	border-bottom:3.0pt solid #0B2338;
	border-left:1.0pt solid #0B2338;
	background:#005B93;
	mso-pattern:auto none;
	vertical-align:top;
	padding-bottom:3.6pt;
	padding-left:7.2pt;
	padding-top:3.6pt;
	padding-right:7.2pt;}
.oa2
	{border-top:3.0pt solid #0B2338;
	border-right:1.0pt solid #0B2338;
	border-bottom:1.0pt solid #0B2338;
	border-left:1.0pt solid #0B2338;
	background:#CBD2DC;
	mso-pattern:auto none;
	vertical-align:top;
	padding-bottom:3.6pt;
	padding-left:7.2pt;
	padding-top:3.6pt;
	padding-right:7.2pt;}
.oa3
	{border:1.0pt solid #0B2338;
	background:#E7EAEE;
	mso-pattern:auto none;
	vertical-align:top;
	padding-bottom:3.6pt;
	padding-left:7.2pt;
	padding-top:3.6pt;
	padding-right:7.2pt;}
.oa4
	{border:1.0pt solid #0B2338;
	background:#CBD2DC;
	mso-pattern:auto none;
	vertical-align:top;
	padding-bottom:3.6pt;
	padding-left:7.2pt;
	padding-top:3.6pt;
	padding-right:7.2pt;}
-->
</style>
</head>

<body>
<!--StartFragment-->

Name | … | States Covered | Ownership | Eligibility Algorithm
-- | -- | -- | -- | --
Medicare |   | * | Government | StandardMedicare
Medicaid |   | * | Government | StandardMedicaid
Dual Eligible |   | * | Government | StandardDualEligible
Private |   | * | Private | Generic
VAHealthCare |   | * | Government | VeteranEligibility

<!--EndFragment-->
</body>

</html>

## Current Eligibility Column Options
_I've bolded the most interesting columns_

- Poverty Multiplier: A basic static federal poverty multiplier income threshold.
- Income Threshold: A basic static income threshold. This is technically redundant due to the qualifying_attributes column, though I think it's worth having as a more user-friendly and common option.
- Age Threshold: A basic age threshold. It is also technically redundant for the same reason as above.
- **Qualifying Codes**: A set of condition codes that would make the person eligible for that algorithm. It can either be input as a set of codes separated by `|` such as `4545454|7878787`. Or you could input a file with a list of eligible codes, such as with `ssd_eligibility.csv`.
- **Qualifying Attributes**: A set of attributes and logical operators that would make the person eligible for that algorithm. You can input it as: `attribute` `operator` `value`. It supports `<,<=,>,>=,=,!=`. It can either be input as a set of attributes separated by `|` such as `pregnant = true|a1c <= .05`. Or you could input a file with a list of eligible attribute logics as in `test_attributes_eligibilities.csv`. It is very particular in that the attribute, operator, and value must be separated by exactly one space. I couldn't find a better solution that works with and without spaces.
- **Poverty Multiplier File**: A file that allows more complex by-state and by-age income multiplier thresholds. An example is `medicaid_eligibility.csv`.
- **Spenddown File**: A file that dictates by-state medical spenddown limits to become eligible for health insurance. One example is Medically Needy Income Limits (MNIL), a Medicaid program that allows non-income eligible people to enroll in Medicaid if they "spend down" on health costs to a certain level. `medicaid_eligibility.csv` includes the by-state Medicaid implementation of this. https://www.medicaid.gov/medicaid/eligibility/index.html#:~:text=Medically%20Needy,-States%20have%20the%20option%20to
- Acceptance Likelihood: A simple factor that dictates the 0.0-1.0 likelihood that a person may be accepted. For example, Social Security Disability may reject up to 10% of eligible applicants, etc.
- **Sub-Eligibilities**: Allows the use of previously defined eligibility algorithms in a new algorithm. To use multiple, separate with `|`.
- **Logical Operator**: A simple operator that dictates whether the eligibility attributes in that algorithm must all be satisfied (and) or just one (or).

Notes:
 - This PR replaces #1010 (Payer Updates) because it includes all commits in that PR.
 - There are lots of files changed in this PR. However, many are trivial changes (one line, method name change, package change, etc.) and several are due to the Medicare/Medicaid eligibilities being split into modular pieces each with its own file.

jawalonoski commented:
@RobertScalfani can you rebase this PR and fix the conflicts?

jawalonoski commented on src/main/resources/payers/insurance_companies.csv:
```diff
@@ -1,10 +1,5 @@
-0,10001,Medicare,7500 Security Blvd,Baltimore,MD,21244,1-800-633-4227,*,*,0,0.8,40,250,Government
-1,10002,Medicaid,7500 Security Blvd,Baltimore,MD,21244,1-877-267-2323,*,*,0,0.95,60,250,Government
-2,10003,Dual Eligible,7500 Security Blvd,Baltimore,MD,21244,1-877-267-2323,*,*,0,0.95,100,500,Government
-3,10004,Humana,500 West Main St,Louisville,KY,40018,1-844-330-7799,*,"wellness,medications,ambulatory,inpatient,emergency,urgentcare",1000,0.75,80,1700,Private
-4,10005,Blue Cross Blue Shield,Michigan Plaza,Chicago,IL,60007,1-800-262-2583,*,*,900,0.75,75,1800,Private
-5,10006,UnitedHealthcare,9800 Healthcare Lane,Minnetonka,MN,55436,1-888-545-5205,*,"wellness,ambulatory,inpatient,emergency,urgentcare,medications",800,0.75,70,1900,Private
-6,10007,Aetna,151 Farmington Ave,Hartford,CT,6156,1-800-872-3862,*,"wellness,ambulatory,inpatient,emergency,urgentcare,medications",700,0.75,65,2000,Private
-7,10008,Cigna Health,900 Cottage Grove Rd,Bloomfield,CT,6002,1-800-997-1654,*,"wellness,medications,ambulatory,inpatient,emergency,urgentcare",600,0.75,60,2100,Private
-8,10009,Anthem,220 Virginia Ave,Indianapolis,IN,46204,1-800-331-1476,*,wellness,500,0.75,55,2200,private
```
Please find a way to reincorporate these large nationwide private payers.

jawalonoski commented on src/main/resources/synthea.properties:
```diff
@@ -225,16 +227,22 @@ generate.providers.minimum = 1
+# The percentage of a person's income that they are willing to spend on health insurance premiums.
+generate.payers.insurance_plans.income_premium_ratio = 0.13
+# The chance of rejection 
 # Payer selection behavior
 # How patients select a payer:
 #  best_rates - select insurance with best rates for person's existing conditions and medical needs
 #  random  - select randomly.
```
The comments list two options: `best_rates` and `random`, yet you've changed the default value to `gov_priority`.

Please update the comments to provide information to our users on the new options they have available.

jawalonoski commented on src/test/resources/generic/payers/test_attributes_eligibility.csv:
```diff
@@ -0,0 +1,8 @@
+attributes
+test1 = true
```
Is `test1 = true` assignment or an equality check?

RobertScalfani replied:
it's an equality check
It's set so a user could use either == or = to try and keep it user friendly.

jawalonoski commented on src/test/java/org/mitre/synthea/world/concepts/LossOfCareHealthRecordTest.java:
```diff
@@ -90,74 +93,80 @@ public void personRunsOutOfIncomeWithNoInsurance() {
+    InsurancePlan plan = testPrivatePayer.getPlans().iterator().next();
+    person.coverage.setPlanAtTime(time, plan);
     person.setProvider(EncounterType.WELLNESS, new Provider());
     Code code = new Code("SNOMED-CT","705129","Fake Code");
-    // Set person's income to be $1 lower than the cost of 8 monthly premiums.
-    person.attributes.put(Person.INCOME, (int) (testPrivatePayer.getMonthlyPremium() * 8) - 1);
+    // Set person's income to be $1 lower than the cost of 7 monthly premiums.
+    person.attributes.put(Person.INCOME, (int) (plan.getMonthlyPremium() * 7) - 1);
 
     // Pay monthly premium 8 times.
-    long oneMonth = Utilities.convertTime("years", 1) / 12;
+    long oneMonth = Utilities.convertTime("months", 1);
```
I vaguely recall this being written with `years / 12` because there is no standard "month" length (28 days to 31 days) and dividing a year by 12 got you closer on the time over many (e.g., 8) months.

Utilities assumes a month of exactly 30 days.

As long as the test still works fine, but over 7 or 8 months this will be 3 to 4 days of difference.

jawalonoski replied:
Oh, this probably also explains why you had to change the income to 7 months instead of 8.

jawalonoski commented on src/test/java/org/mitre/synthea/world/concepts/LossOfCareHealthRecordTest.java:
```diff
@@ -174,7 +183,8 @@ public void personRunsOutOfIncomeDueToMonthlyPremium() {
 
     Person person = new Person(0L);
-    person.coverage.setPayerAtTime(time, Payer.noInsurance);
+    person.attributes.put(Person.BIRTHDATE, time);
```
The birthdate is set again in 6 and 7 lines.

RobertScalfani commented:
Got the merge conflicts resolved. However, I do have some concerns about the Claim costs logic, particularly assignCosts() in Claim.java.

I previously found that many of the cost related tests in PayerTest.java seemed to be flawed because they were inconsistently treating what the different claim attributes did (deductible, coinsurance, pocket, payer, etc.). So I renamed the claim attributes in an attempt to improve that. I also found that there were lots of post-processing on the claim's attributes, so I updated it to pre-process those and keep them in new attributes (coinsurancePaidByPatient, coinsurancePaidByPayer, etc.). I also updated the assignCosts() logic itself in an attempt to remove nested if's and make it more readable. It was just difficult for me to parse through it when debugging and updating the failing costs tests.

Having said that, I'm concerned that this new logic could be wrong/incomplete because of my lack of knowledge regarding claims. I know lots of claims updates have been made to Synthea and I don't want to be stepping on and overwriting anything. I am particularly concerned with how those claim attribute are dealt with in the BB2RIFExporter.java.

Making note of this so we can track it and look deeper at it.

It may turn out that it would be just easier to revert this back to how it was before, though there were those odd test inconsistencies that would need to be resolved. I'll look deeper into this once I'm back on the 26th.

eedrummer commented on src/main/java/org/mitre/synthea/helpers/Utilities.java:
```diff
@@ -115,6 +115,15 @@ public static int getMonth(long time) {
+  /**
+   * Get the the number of days from a Unix timestamp.
+   */
+  public static double getDurationDays(long time) {
```
This method does not behave the way that I would expect. Given that it returns a `double`, I expect that it would return fractions of days. When passing in the value `43200000`, it returns `0.0`, not `0.5` as I would expect.

I recommend either changing the return type to `long` if this is the intended behavior. If not, dividing two longs will provide a whole number. Casting it to a `double` after this division will not provide additional precision.

RobertScalfani replied:
resolved by using BigDecimals

eedrummer replied:
Is there code that still needs to be checked in? I still see it using `long` values.

eedrummer replied:
Sorry.. somehow was looking at an older commit.

eedrummer commented on src/main/java/org/mitre/synthea/world/agents/behaviors/planeligibility/CSVEligibility.java:
```diff
@@ -0,0 +1,118 @@
+
+public class CSVEligibility implements IPlanEligibility {
+
+  // The possible columns of the CSV input file.
```
I think there is a little bit of a user interface issue to discuss here.

The users of CSV Eligibility files are going to be mostly non-developers. However, column names are in snake case. This isn't a big deal, but it is worth considering if these should be in just normal title case or if the code should be a bit more robust to things like capitalizing parts or all of the column names.

RobertScalfani replied:
I've updated the csv eligibility columns to use white spaces and capitalizations... As a larger approach, though, should this same format be extended to Payers and Plans?

eedrummer commented on src/main/java/org/mitre/synthea/world/agents/behaviors/planeligibility/CSVEligibility.java:
```diff
@@ -0,0 +1,118 @@
+   * @param logicalOperator The input string.
+   * @return  The converted logical operator.
+   */
+  private static String convertToLogicalOperator(String logicalOperator) {
```
Related to an above comment, you should probably trim whitespace on these to avoid throwing an exception that will be hard for people to understand. Essentially the error message would just have an extra whitespace in it.

eedrummer commented on src/main/java/org/mitre/synthea/world/agents/behaviors/planeligibility/GenericPayerEligibilty.java:
```diff
@@ -0,0 +1,16 @@
+/**
+ * A payer eligibilty algorithm that accepts all patients.
+ */
+public class GenericPayerEligibilty implements IPlanEligibility {
```
This class is only instantiated once. Given that it is an implementation of a single method interface, it can just be replaced with a lambda where it gets instantiated.

RobertScalfani replied:
I was able to do this for GenericPayerEligibility, but not for any of the others you pointed out since they don't use the default constructor and have fields.

eedrummer commented on src/main/java/org/mitre/synthea/world/agents/behaviors/planeligibility/IncomeThresholdEligibility.java:
```diff
@@ -0,0 +1,19 @@
+
+import org.mitre.synthea.world.agents.Person;
+
+public class IncomeThresholdEligibility implements IPlanEligibility {
```
See comment on `GenericPayerEligibilty`

eedrummer commented on src/main/java/org/mitre/synthea/world/agents/behaviors/planeligibility/MedicallyNeedyIncomeEligibility.java:
```diff
@@ -0,0 +1,69 @@
+    while (csv.hasNext()) {
+      Map<String, String> row = csv.next();
+      if (row.get("state").equals(state)) {
+        mnilAvailable = Boolean.parseBoolean(row.get("mnil-available"));
```
Column names should be constants like you have them in other classes.

eedrummer commented on src/main/java/org/mitre/synthea/world/agents/behaviors/planeligibility/PovertyMultiplierEligibility.java:
```diff
@@ -0,0 +1,21 @@
+import org.mitre.synthea.modules.HealthInsuranceModule;
+import org.mitre.synthea.world.agents.Person;
+
+public class PovertyMultiplierEligibility implements IPlanEligibility {
```
Again, see comment on `GenericPayerEligibilty`

eedrummer commented on src/main/java/org/mitre/synthea/world/agents/behaviors/planeligibility/QualifyingAttributesEligibility.java:
```diff
@@ -0,0 +1,154 @@
+      resource = Utilities.readResource(fileName);
+      csv = SimpleCSV.parseLineByLine(resource);
+    } catch (IOException e) {
+      e.printStackTrace();
```
Again, you'll want to do more than print the stacktrace.

eedrummer commented on src/main/java/org/mitre/synthea/world/agents/behaviors/planeligibility/QualifyingAttributesEligibility.java:
```diff
@@ -0,0 +1,154 @@
+
+  private static AttributeQualifier covertAttributeExpressionToLogic(String attributeExpression) {
+    String regex = " ";
+    String[] splitExpression = attributeExpression.split(regex);
```
This is another place to think about the user experience. The expression has to be exactly attribute, one space character, comparator, one space character and then value.

Users may expect to be able to omit the whitespace, or include more whitespace without it being significant. Consider a more robust regex with match groups to allow more variation in input.

eedrummer commented on src/main/java/org/mitre/synthea/world/agents/behaviors/planfinder/PlanFinderGovPriority.java:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    if (eligiblePlans.stream().anyMatch(plan
+        -> plan.getPayer().getName().equals(PayerManager.DUAL_ELIGIBLE))) {
```
Seems like it would be handy to have an isDuelEligible method.

eedrummer commented on src/main/java/org/mitre/synthea/world/agents/PayerManager.java:
```diff
@@ -0,0 +1,388 @@
+  private static IPlanFinder planFinder;
+  // Payer selection algorithm choices:
+  private static final String RANDOM = "random";
+  private static final String BESTRATE = "best_rate";
```
This is really nit-picky, but suggest that you are consistent in constant naming. The rest of the file uses underscores to replace spaces as opposed to omitting them.

eedrummer commented:
Hey @RobertScalfani - Jay and I provided comments throughout. Let us know when you can get these addressed and when we should take another pass at reviewing.

eedrummer reviewed (changes requested):
This is getting close. There are a few things that should be addressed.

First, you should add a page to the wiki, describing the feature and providing an example. You can start the page now, but leave a note at the top that it is a WIP on the `detailed-finances` branch.

Second, there are two issues in the code that need fixing.

The checkstyle report is flagging some test code. I'm guessing that it isn't something you touched, but maybe you need to sync your branch with the latest on `master`?

I'm also getting an exception for one case when running the code. If I run:
```
./run_synthea -p 1000 -s 1659721109702 -cs 1659721109702
```

I get:
```
462 -- Glenn Goodwin (54 y/o F) Holden, Massachusetts  (76283)
java.lang.RuntimeException: Person does not have insurance for age 0.
        at org.mitre.synthea.world.agents.Person.incomeRemaining(Person.java:868)
        at org.mitre.synthea.world.agents.behaviors.planeligibility.IncomeSpenddownEligibility.isPersonEligible(IncomeSpenddownEligibility.java:70)
        at org.mitre.synthea.world.agents.behaviors.planeligibility.CSVEligibility.lambda$isPersonEligible$2(CSVEligibility.java:75)
        at java.base/java.util.stream.MatchOps$1MatchSink.accept(MatchOps.java:90)
        at java.base/java.util.ArrayList$ArrayListSpliterator.tryAdvance(ArrayList.java:1602)
        at java.base/java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:129)
        at java.base/java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:527)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:513)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
...
```
I saw this the first time just trying to generate 1000 patients. I was able to reproduce the issue using the seed and clinician seed.

I'm seeing this on macOS
```
❯ java --version                                             
openjdk 17.0.3 2022-04-19
OpenJDK Runtime Environment Temurin-17.0.3+7 (build 17.0.3+7)
OpenJDK 64-Bit Server VM Temurin-17.0.3+7 (build 17.0.3+7, mixed mode)
```

jawalonoski commented:
Reviewing this now... reminder that you should **ALWAYS** use `git mv` when you want to move files to new or different packages. Just moving the file to different folders without using `git mv` destroys the history of that file. This means all our behavior classes lose years of commit history.

jawalonoski commented:
> Reviewing this now... reminder that you should **ALWAYS** use `git mv` when you want to move files to new or different packages. Just moving the file to different folders without using `git mv` destroys the history of that file. This means all our behavior classes lose years of commit history.

Actually, it looks like GitHub webpages preserves this... but strangely, my git client doesn't recognize it and show me the history. 🤔

jawalonoski commented on src/main/java/org/mitre/synthea/export/CSVExporter.java:
```diff
@@ -480,7 +510,9 @@ public void export(Person person, long time) throws IOException {
       for (Medication medication : encounter.medications) {
         medication(personID, encounterID, payerID, medication, time);
-        claim(person, medication.claim, encounter, encounterID, time);
+        if (Config.getAsBoolean("exporter.csv.claim_transactions")) {
+          claim(person, medication.claim, encounter, encounterID, time);
+        }
```
Not sure about these `if` code blocks and the two new parameters... if you don't want those files, then intent was to for you to use the `exporter.csv.excluded_files` property. For example,

```properties
# included_files and excluded_files list out the files to include/exclude in the csv exporter
# only one of these may be set at a time, if both are set then both will be ignored
# if neither is set, then all files will be included
# see list of files at: https://github.com/synthetichealth/synthea/wiki/CSV-File-Data-Dictionary
# include filenames separated with a comma, ex: patients.csv,procedures.csv,medications.csv
# NOTE: the csv exporter does not actively delete files, so if Run 1 you included a file, then Run 2 you exclude that file, the version from Run 1 will still be present
exporter.csv.included_files =
exporter.csv.excluded_files = claims.csv,claims_transactions.csv,patient_expenses.csv
```


I suppose using `exporter.csv.claim_transactions` and `exporter.csv.patient_expenses` is marginally faster, but it is special casing the claims processing.

jawalonoski reviewed (commented):
Still reviewing... apologies for the delay.

RobertScalfani commented:
> Reviewing this now... reminder that you should **ALWAYS** use `git mv` when you want to move files to new or different packages. Just moving the file to different folders without using `git mv` destroys the history of that file. This means all our behavior classes lose years of commit history.

Good to know, thanks. Will do in future

RobertScalfani commented on src/main/java/org/mitre/synthea/world/agents/behaviors/planeligibility/QualifyingAttributesEligibility.java:
```diff
@@ -0,0 +1,159 @@
+    return qualifyingAttributes;
+  }
+
+  private static AttributeQualifier covertAttributeExpressionToLogic(String attributeExpression) {
```
😂

jawalonoski commented on src/test/java/org/mitre/synthea/engine/StateTest.java:
```diff
@@ -85,12 +86,9 @@ public void setup() throws IOException {
-      person.coverage.setPayerAtTime(yearTime, Payer.noInsurance);
-    }
-
+    PayerManager.loadPayers(new Location("Massachusetts", null));
```
Ideally, you do not want to specify a location in the unit tests like "Massachusetts" because it will break synthea-international support. Instead, you want to do something like

```
String testStateDefault = Config.get("test_state.default", "Massachusetts");
PayerManager.loadPayers(new Location(testStateDefault, null));
```

jawalonoski commented on src/test/java/org/mitre/synthea/engine/LogicTest.java:
```diff
@@ -63,8 +64,9 @@ public void setup() throws IOException {
     // Ensure Person's Payer is not null.
-    Payer.loadNoInsurance();
-    person.coverage.setPayerAtTime(time, Payer.noInsurance);
+    PayerManager.loadPayers(new Location("Massachusetts", null));
```
Ideally, you do not want to specify a location in the unit tests like "Massachusetts" because it will break synthea-international support. Instead, you want to do something like

```
String testStateDefault = Config.get("test_state.default", "Massachusetts");
PayerManager.loadPayers(new Location(testStateDefault, null));
```

jawalonoski commented on src/test/java/org/mitre/synthea/export/ExporterTest.java:
```diff
@@ -52,11 +52,9 @@ public void setup() throws Exception {
-      long yearTime = time - years(i);
-      patient.coverage.setPayerAtTime(yearTime, Payer.noInsurance);
-    }
+    PayerManager.loadPayers(new Location("Massachusetts", null));
```
Ideally, you do not want to specify a location in the unit tests like "Massachusetts" because it will break synthea-international support. Instead, you want to do something like

```
String testStateDefault = Config.get("test_state.default", "Massachusetts");
PayerManager.loadPayers(new Location(testStateDefault, null));
```

jawalonoski commented on src/test/java/org/mitre/synthea/export/FHIRDSTU2ExporterTest.java:
```diff
@@ -239,12 +241,9 @@ public void testObservationAttachment() throws Exception {
-      long yearTime = time - Utilities.convertTime("years", i);
-      person.coverage.setPayerAtTime(yearTime, Payer.noInsurance);
-    }
+    PayerManager.loadPayers(new Location("Massachusetts", null));
```
Ideally, you do not want to specify a location in the unit tests like "Massachusetts" because it will break synthea-international support. Instead, you want to do something like

```
String testStateDefault = Config.get("test_state.default", "Massachusetts");
PayerManager.loadPayers(new Location(testStateDefault, null));
```

jawalonoski commented on src/test/java/org/mitre/synthea/export/FHIRDSTU2ExporterTest.java:
```diff
@@ -62,6 +64,10 @@ public static void setup() {
     // Ensure Physiology state is enabled
     physStateEnabled = State.ENABLE_PHYSIOLOGY_STATE;
     State.ENABLE_PHYSIOLOGY_STATE = true;
+    // Ensure plan and eligibilities are loaded.
+    PlanEligibilityFinder.buildPlanEligibilities(Config.get("test_state.default", "Massachusetts"),
+        Config.get("generate.payers.insurance_plans.eligibilities_file"));
```
👍

jawalonoski commented on src/test/java/org/mitre/synthea/export/FHIRR4ExporterTest.java:
```diff
@@ -62,6 +62,7 @@ public static void setup() {
     // Ensure Physiology state is enabled
     physStateEnabled = State.ENABLE_PHYSIOLOGY_STATE;
     State.ENABLE_PHYSIOLOGY_STATE = true;
+    PayerManager.loadPayers(new Location("Massachusetts", null));
```
Ideally, you do not want to specify a location in the unit tests like "Massachusetts" because it will break synthea-international support. Instead, you want to do something like

```
String testStateDefault = Config.get("test_state.default", "Massachusetts");
PayerManager.loadPayers(new Location(testStateDefault, null));
```

jawalonoski commented on src/test/java/org/mitre/synthea/export/FHIRSTU3ExporterTest.java:
```diff
@@ -64,6 +65,7 @@ public static void setup() {
     // Ensure Physiology state is enabled
     physStateEnabled = State.ENABLE_PHYSIOLOGY_STATE;
     State.ENABLE_PHYSIOLOGY_STATE = true;
+    PayerManager.loadPayers(new Location("Massachusetts", null));
```
Ideally, you do not want to specify a location in the unit tests like "Massachusetts" because it will break synthea-international support. Instead, you want to do something like

```
String testStateDefault = Config.get("test_state.default", "Massachusetts");
PayerManager.loadPayers(new Location(testStateDefault, null));
```

jawalonoski commented on src/test/java/org/mitre/synthea/modules/DeathModuleTest.java:
```diff
@@ -44,8 +46,9 @@ public void setup() throws IOException {
     person.attributes.put(Person.BIRTHDATE, birthTime);
-    Payer.loadNoInsurance();
-    person.coverage.setPayerAtTime(time, Payer.noInsurance);
+    PayerManager.loadPayers(new Location("Massachusetts", null));
```
Ideally, you do not want to specify a location in the unit tests like "Massachusetts" because it will break synthea-international support. Instead, you want to do something like

```
String testStateDefault = Config.get("test_state.default", "Massachusetts");
PayerManager.loadPayers(new Location(testStateDefault, null));
```

jawalonoski commented on src/test/java/org/mitre/synthea/modules/EncounterModuleTest.java:
```diff
@@ -38,8 +38,9 @@ public static void setup() throws Exception {
     // Ensure Person's Payer is not null.
-    Payer.loadNoInsurance();
-    person.coverage.setPayerAtTime(System.currentTimeMillis(), Payer.noInsurance);
+    PayerManager.loadPayers(new Location("Massachusetts", null));
```
Ideally, you do not want to specify a location in the unit tests like "Massachusetts" because it will break synthea-international support. Instead, you want to do something like

```
String testStateDefault = Config.get("test_state.default", "Massachusetts");
PayerManager.loadPayers(new Location(testStateDefault, null));
```

jawalonoski commented on src/test/java/org/mitre/synthea/world/agents/PayerTest.java:
```diff
@@ -54,163 +56,237 @@ public static void setup() throws Exception {
     // Set up Medicaid numbers.
     healthInsuranceModule = new HealthInsuranceModule();
-    double povertyLevel =
+    povertyLevel =
             Config.getAsDouble("generate.demographics.socioeconomic.income.poverty", 11000);
```
You changed this default value from `11000` to `12880` in `synthea.properties` so if the `12880` was intentional you may want to make this consistent.

jawalonoski commented on src/main/java/org/mitre/synthea/world/agents/behaviors/planeligibility/IPlanEligibility.java:
```diff
@@ -0,0 +1,19 @@
+   * algorithm at the given time.
+   * @param person The person to check against.
+   * @param time   The time.
+   * @return Whether the person is elgible.
```
```suggestion
   * @return Whether the person is eligible.
```

jawalonoski reviewed (commented):
I think this is my final set of feedback.

All of my comments about `Config.get("test_state.default", "Massachusetts")` might possibly be able to use `Generator.DEFAULT_STATE` if it was previously set... not sure about that though. Just thought I'd throw that out there.

RobertScalfani commented on src/main/java/org/mitre/synthea/world/concepts/healthinsurance/InsurancePlan.java:
```diff
@@ -0,0 +1,266 @@
+    return (service == null
+        || this.servicesCovered.contains(service)
+        || this.servicesCovered.contains("*"))
+        && this.payer.isInNetwork(null);
```
Removed the isInNetwork() method, since it's not implemented yet it shouldn't be a part of this PR.

jawalonoski commented:
@eedrummer this PR looks good to me... were there any other outstanding changes you wanted to see?

eedrummer commented:
I'm seeing two issues:

* There is a whitespace check style rule violation that should be fixed
* I'm getting a repeatable test failure, but I realize that it seems to be building fine in CI. 

Here is the failure I keep getting:

```
org.mitre.synthea.world.agents.PayerTest > receiveMedicareAgeEligible FAILED
    java.lang.RuntimeException: Person does not have insurance for age P46Y11M24D at time -726451200000.
        at org.mitre.synthea.world.concepts.healthinsurance.CoverageRecord.incomeRemaining(CoverageRecord.java:373)
        at org.mitre.synthea.world.agents.behaviors.planeligibility.IncomeSpenddownEligibility.isPersonEligible(IncomeSpenddownEligibility.java:73)
        at org.mitre.synthea.world.agents.behaviors.planeligibility.CSVEligibility.lambda$convertToLogic$9(CSVEligibility.java:127)
        at java.base/java.util.stream.MatchOps$1MatchSink.accept(MatchOps.java:90)
        at java.base/java.util.ArrayList$ArrayListSpliterator.tryAdvance(ArrayList.java:1602)
        at java.base/java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:129)
        at java.base/java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:527)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:513)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
        at java.base/java.util.stream.MatchOps$MatchOp.evaluateSequential(MatchOps.java:230)
        at java.base/java.util.stream.MatchOps$MatchOp.evaluateSequential(MatchOps.java:196)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.base/java.util.stream.ReferencePipeline.anyMatch(ReferencePipeline.java:632)
        at org.mitre.synthea.world.agents.behaviors.planeligibility.CSVEligibility.lambda$convertToLogic$9def446d$1(CSVEligibility.java:126)
        at org.mitre.synthea.world.agents.behaviors.planeligibility.CSVEligibility.isPersonEligible(CSVEligibility.java:67)
        at org.mitre.synthea.world.agents.behaviors.planeligibility.CSVEligibility.lambda$convertToLogic$10(CSVEligibility.java:132)
        at java.base/java.util.stream.MatchOps$1MatchSink.accept(MatchOps.java:90)
        at java.base/java.util.ArrayList$ArrayListSpliterator.tryAdvance(ArrayList.java:1602)
        at java.base/java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:129)
        at java.base/java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:527)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:513)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
        at java.base/java.util.stream.MatchOps$MatchOp.evaluateSequential(MatchOps.java:230)
        at java.base/java.util.stream.MatchOps$MatchOp.evaluateSequential(MatchOps.java:196)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.base/java.util.stream.ReferencePipeline.allMatch(ReferencePipeline.java:637)
        at org.mitre.synthea.world.agents.behaviors.planeligibility.CSVEligibility.lambda$convertToLogic$9def446d$2(CSVEligibility.java:131)
        at org.mitre.synthea.world.agents.behaviors.planeligibility.CSVEligibility.isPersonEligible(CSVEligibility.java:67)
        at org.mitre.synthea.world.concepts.healthinsurance.InsurancePlan.accepts(InsurancePlan.java:238)
        at org.mitre.synthea.world.agents.behaviors.planfinder.PlanFinderPriority.lambda$find$0(PlanFinderPriority.java:21)
        at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:178)
        at java.base/java.util.HashMap$KeySpliterator.forEachRemaining(HashMap.java:1707)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
        at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
        at org.mitre.synthea.world.agents.behaviors.planfinder.PlanFinderPriority.find(PlanFinderPriority.java:23)
        at org.mitre.synthea.world.agents.PayerManager.findPlan(PayerManager.java:377)
        at org.mitre.synthea.modules.HealthInsuranceModule.updateInsurance(HealthInsuranceModule.java:79)
        at org.mitre.synthea.modules.HealthInsuranceModule.process(HealthInsuranceModule.java:61)
        at org.mitre.synthea.world.agents.PayerTest.receiveMedicareAgeEligible(PayerTest.java:220)
```

RobertScalfani commented:
Following up on this

RobertScalfani commented:
Resolved
